core/arch/arm/mm/pgt_cache.c

```diff
@@ -244,15 +265,7 @@ static void pgt_free_unlocked(struct pgt_cache *pgt_cache)
 		struct pgt *p = SLIST_FIRST(pgt_cache);
 
 		SLIST_REMOVE_HEAD(pgt_cache, link);
-		if (p->num_used_entries) {
-			push_to_cache_list(p);
-		} else {
-			tee_pager_pgt_save_and_release_entries(p);
-			p->ctx = NULL;
-			p->vabase = 0;
-
-			push_to_free_list(p);
-		}
+		push_to_cache_list(p);
```

Comment:
here pager (with 32bit mmu) seems to treat unused pgts differently eventually releasing rw paged memory. Can the situation happen here (`p->num_used_entries==0`)?

Reply:
Before this, it only made sense to cache pgt's where the pager had pages assigned. Now we're going to use this as a way of avoiding to reinitialized the translation tables where possible.

Reply:
I agree when pager is off, caching pgt's allows to possibly reuse them, but when pager is enable, we don't have much secure RAM and releasing secure memory can be quite useful for smooth optee execution. Pgt's are pager locked rw pages. With this change they remain paged until the TA context is released. This could prevent paged op-tee from executing several TAs  or make pager very inefficient because of limited number of available physical pages in the pager pool.

Reply:
They can still be re-used, it's just that they need to be picked up from `pgt_cache_list` instead of `pgt_free_list`. `pop_from_some_list()` handles this.

Reply:
The previous implementation released the pgt's rw locked pages once TA invocation completes. With the new implementation (this P-R), the pgt's rw locked pages are still in place, in pgt cache list, until the TA context dies, hence lowering the number of available physical pages for the service next invocation. It is a matter of balance I think on, when pager is enable, whether we prefer to speed up time to setup back TA mapping, or spend more time but relax the pager pool pressure.

I've tested this P-R against `xtest regression_1005` that attempts the created many sessions on a multi instance TA. I've increased the number of opened sessions from 3 to 10. I've compared the behavior with current change and with the hack below to restore release of pager locked pgt's from `pgt_free_unlocked()`:
```patch
 	while (!SLIST_EMPTY(pgt_cache)) {
 		struct pgt *p = SLIST_FIRST(pgt_cache);
 
 		SLIST_REMOVE_HEAD(pgt_cache, link);
 
+#ifdef CFG_PAGED_USER_TA
+		if (p->num_used_entries) {
+			/* Release paged locked PGT's now from pager */
+			tee_pager_pgt_save_and_release_entries(p);
+			p->ctx = NULL;
+			p->vabase = 0;
+
+			push_to_free_list(p);
+			continue;
+		}
+#endif
 		push_to_cache_list(p);
 	}
```
On this very test and on my paged platform, releasing pgt's on TA context destruction (the P-R as-is) rather than on TA invocation completion (P-R + the above patch) makes this test about 25% slower. Tracking the number of unlocked pages in page pool, I see its a it about 20% higher with the above patch applied (because pgts are released).

I am thinking also of keep alive TAs that, IIUC, will have their cache pgt's assigned forever, preventing pager from using the related rw lock pages.

Maybe I missing something but high pressure on pager pool can make OP-TEE very very slow to operate.

Reply:
Thanks, I see the problem now.

Reply:
In your diff above you're freeing pages more aggressively (all physical pages referred by this table too) than before this PR.

I've added an update restoring the old behavior so it's slightly different from what you suggested above.